dbg.h

```diff
@@ -298,8 +303,10 @@ class DebugOutput {
 
 // We use a variadic macro to support commas inside expressions (e.g.
 // initializer lists):
-#define dbg(...)                                                     \
-  dbg_macro::DebugOutput(__FILE__, __LINE__, __func__, #__VA_ARGS__) \
-      .print((__VA_ARGS__))
+#define dbg(...)                                                              \
+  dbg_macro::DebugOutput(__FILE__, __LINE__, __func__, #__VA_ARGS__)          \
+      .print(                                                                 \
+          dbg_macro::type_name(dbg_macro::type_tag<decltype(__VA_ARGS__)>{}), \
```

Comment:
Using `decltype(__VA_ARGS__)` preserves the actual type of the expressions and so
```cpp
int const &i = a;
dbg(i);
```
will actually output `const int &` whereas previously it would have just printed `int`.

Reply:
The downside is that you need overloads for all combinations of cv- and ref-qualifiers, i.e.
```cpp
std::string type_name(type_tag<std::string>)   { return "std::string"; }
std::string type_name(type_tag<std::string&>)  { return "std::string&"; }
std::string type_name(type_tag<std::string&&>) { return "std::string&&"; }
std::string type_name(type_tag<const std::string>)   { return "const std::string"; }
std::string type_name(type_tag<const std::string&>)  { return "const std::string&"; }
std::string type_name(type_tag<const std::string&&>) { return "const std::string&&"; }
std::string type_name(type_tag<volatile std::string>)   { return "volatile std::string"; }
std::string type_name(type_tag<volatile std::string&>)  { return "volatile std::string&"; }
std::string type_name(type_tag<volatile std::string&&>) { return "volatile std::string&&"; }
std::string type_name(type_tag<const volatile std::string>)   { return "const volatile std::string"; }
std::string type_name(type_tag<const volatile std::string&>)  { return "const volatile std::string&"; }
std::string type_name(type_tag<const volatile std::string&&>) { return "const volatile std::string&&"; }
```
I guess one could come up with some compiler magic for that as well, but it's questionable as to how far to hold the user's hand.

Reply:
Oh and the overloads for pointers are not even covered yet.